### ISSUE TITLE
RFD: enrich claims for OIDC and SAML

### ISSUE THREAD

**Issue by klizhentas** [gravitational/teleport#6707]
## Details

Many times admins don't have access to OIDC/SAML identity providers. Teleport should provide a tool similar to Auth0 rules that allows to modify and code claims on the fly.

The approach below is fully declarative, users can define new traits as a code and they can reduce reliance on identity provider. Users can define all claims through teleport, reducing using the identity provider to a simple email database.

## Login rules

We will decouple user's identity from claims we inject. A new resource `login_rule` will define a rule that will be processed right after the OIDC or SAML callback. This rule will be applied to ephemeral user resource that is generated after SAML/OIDC callback.

```yaml
type: login_rule
spec: 
   match:
       user: 'alice@example.com'
   rewrite:
      # sets attribute 'groups': ['admins'] for user alice
      'groups': ['admins']
```

## Removing attributes

Admins can define dynamic rules to add or remove some traits conditionally or unconditionally.

```yaml
type: login_rule
spec:
    match:
        where: contains(user.metadata.traits, 'admin')
   # no admins allowed
    rewrite:
    - remove(user.groups, 'admin')
```

## Rewriting username

Admins can rewrite usernames based on external claims and rules.

```yaml
type: login_rule
spec:
    match:
        where: '*'
    # rewrite user resource from custom trait 'custom-group' instead of email
    rewrite:
    - set(user.metadata.name, user.metadata.traits['custom-group'])
```

### RBAC 

Teams can have limited scope to add/remove attributes for their sub-teams.

```yaml
type: role
spec:
   allow:
      resources: [login_rules]
      - login_rules
      where: # match label
```

## Export claims from third party services

A customer 'acme' has a third-party storage of information about users stored in a third-party database. Build a plugin that syncs claims from postgres database and adds login rules for each user.

Unlike "access request" approach with waiting room, it does not stand in the way of the login flow and will work even if plugins are down.

## Terraform plugin

Add login to terraform plugin to allow users to define the rules declaratively, as a code.

**Comment by klizhentas** [comment]
Early thoughts:

We can decouple user's identity with claims we inject. For example, this could be a resource:

type: rule
spec: 
   match:
       user: alice@example.com
   rewrite_claims:
      'groups': ['admins']

Rules could be more sophisticated:

type: rule
spec:
    match:
        where: contains(user.groups, 'admin')
   # no admins allowed
    rewrite_claims:
    - remove(user.groups, 'admin')

**Comment by Aharic** [comment]
Would be good to have the rewrite option extend to things like usernames as well (per customer request) so that admins can have a user inherit a username from the IdP and then append to it or rewrite it in full to match their teleport-specific scheme. 

For example: if username passed back by ADFS is `user`, an admin would like to have the ability to rewrite it to something like 'user-admin' or 'user-devops' depending on their needs internally.

**Comment by klizhentas** [comment]
Sometimes additional data is contained in another service (e.g. API). Find a way to add remove or update claims using GRPC plugin or service.

**Comment by klizhentas** [comment]
We can create a system similar to access requests. If the mode is activated, the system will create a `login_request'. The plugin will watch login requests and approve or deny them. The plugin can also add arbitrary claims as well. We can make it easy to add claims using our typescript plugin system.

**Comment by zmb3** [comment]
@nklaassen assigning this to you as I believe your trait transforms work should address the spirit of this issue.

**Comment by ptgott** [comment]
Removing the `documentation` label as this looks like a development ticket. Once we have a plan to get this done, though, let the Docs Team know and we can get the docs work scheduled!

**Comment by nklaassen** [comment]
Login Rules should address this, RFD is in progress here https://github.com/gravitational/teleport/pull/14977

**Comment by nklaassen** [comment]
This is now addressed by Login Rules released in v11.3.0, draft docs are posted here: https://github.com/gravitational/teleport/pull/21195
